src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -231,101 +216,164 @@ private static bool CanOptimizeSingleSpreadAsCollectionBuilderArgument(BoundColl
+                }
 
-                if (elements.Length == 0)
+                const WellKnownMember wellKnownMember = WellKnownMember.System_Runtime_InteropServices_ImmutableCollectionsMarshal__AsImmutableArray_T;
+                if (_compilation.GetWellKnownTypeMember(wellKnownMember) is not MethodSymbol asImmutableArray)
+                {
+                    // ReadOnlySpan<T> span = [elems];
+                    // ImmutableArray<T> array = ImmutableArray.Create(span)
+                    return VisitCollectionBuilderCollectionExpression(node);
+                }
+
+                // T[] array = [elems];
+                // ImmutableArray<T> value = ImmutableCollectionsMarshal.AsImmutableArray(array);
+                var arrayType = getBackingArrayType();
+                var arrayValue = createArray(arrayType, targetsReadOnlyCollection: true);
+                return _factory.StaticCall(asImmutableArray.Construct([arrayType.ElementTypeWithAnnotations]), [arrayValue]);
+            }
+
+            BoundExpression? tryCreateNonArrayBackedSpan(bool isReadOnlySpan)
+            {
+                Debug.Assert(isReadOnlySpan
+                    ? collectionType.OriginalDefinition == (object)_compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T)
+                    : collectionType.OriginalDefinition == (object)_compilation.GetWellKnownType(WellKnownType.System_Span_T));
+
+                var spanType = (NamedTypeSymbol)collectionType;
+                var elementType = spanType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0];
+                if (node.Elements.Length == 0)
                 {
                     // `default(Span<T>)` is the best way to make empty Spans
                     return _factory.Default(collectionType);
                 }
 
-                if (collectionTypeKind == CollectionExpressionTypeKind.ReadOnlySpan &&
+                if (isReadOnlySpan &&
                     ShouldUseRuntimeHelpersCreateSpan(node, elementType.Type))
                 {
                     // Assert that binding layer agrees with lowering layer about whether this collection-expr will allocate.
-                    Debug.Assert(!IsAllocatingRefStructCollectionExpression(node, collectionTypeKind, elementType.Type, _compilation));
+                    Debug.Assert(!IsAllocatingRefStructCollectionExpression(node, CollectionExpressionTypeKind.ReadOnlySpan, elementType.Type, _compilation));
                     var constructor = ((MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_ReadOnlySpan_T__ctor_Array)).AsMember(spanType);
-                    var rewrittenElements = elements.SelectAsArray(static (element, rewriter) => rewriter.VisitExpression((BoundExpression)element), this);
+                    var rewrittenElements = node.Elements.SelectAsArray(static (element, rewriter) => rewriter.VisitExpression((BoundExpression)element), this);
```

Comment:
The local function is named "tryCreateNonArrayBackedSpan", but we're creating an array-backed span here